common/cuda_hip/base/math.hpp

```diff
@@ -11,16 +11,57 @@
 #include <ginkgo/core/base/math.hpp>
 
 
+#ifdef GKO_COMPILING_CUDA
+
+
+#include <cuda_fp16.h>
+
+
+#elif defined(GKO_COMPILING_HIP)
+
+
+#include <hip/hip_fp16.h>
+
+
+#endif
+
+
 namespace gko {
 
 
 // We need this struct, because otherwise we would call a __host__ function in a
 // __device__ function (even though it is constexpr)
 template <typename T>
 struct device_numeric_limits {
-    static constexpr auto inf = std::numeric_limits<T>::infinity();
-    static constexpr auto max = std::numeric_limits<T>::max();
-    static constexpr auto min = std::numeric_limits<T>::min();
+    static constexpr auto inf() { return std::numeric_limits<T>::infinity(); }
+    static constexpr auto max() { return std::numeric_limits<T>::max(); }
+    static constexpr auto min() { return std::numeric_limits<T>::min(); }
```

Comment:
`__half` does not have constexpr constructor such that I can not put them into static constexpr data here.
I do not try the following such that keep them as data yet.
If we still make it as the data not functions, the definition is on the host.
Can cuda still use them in device code? I doubt not because we do not pass it through the kernel call.

Thus, I made them into functions.